include/xtensor-io/xio_zlib.hpp

```diff
@@ -0,0 +1,347 @@
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XTENSOR_IO_ZLIB_HPP
+#define XTENSOR_IO_ZLIB_HPP
+
+#include <fstream>
+
+#include "zlib.h"
+
+#include "xtensor/xadapt.hpp"
+#include "xtensor-io.hpp"
+#include "xfile_array.hpp"
+#include "xio_stream_wrapper.hpp"
+
+#define ZLIB_CHUNK 0x4000
+
+namespace xt
+{
+    namespace detail
+    {
+        inline std::string zlib_err(int ret)
+        {
+            switch (ret)
+            {
+                case Z_STREAM_ERROR:
+                    return "invalid compression level";
+                case Z_DATA_ERROR:
+                    return "invalid or incomplete deflate data";
+                case Z_MEM_ERROR:
+                    return "out of memory";
+                case Z_VERSION_ERROR:
+                    return "zlib version mismatch";
+                default:
+                    return "unknown";
+            }
+        }
+
+        template <typename T, class I>
+        inline xt::svector<T> load_zlib(I& stream, bool as_big_endian)
+        {
+            xt::svector<T> uncompressed_buffer;
+            int ret;
+            unsigned have;
+            z_stream strm;
+            unsigned char in[ZLIB_CHUNK];
+            T out[ZLIB_CHUNK / sizeof(T)];
+
+            strm.zalloc = Z_NULL;
+            strm.zfree = Z_NULL;
+            strm.opaque = Z_NULL;
+            strm.avail_in = 0;
+            strm.next_in = Z_NULL;
+            ret = inflateInit(&strm);
+            if (ret != Z_OK)
+            {
+                XTENSOR_THROW(std::runtime_error, "zlib decompression failed (" + zlib_err(ret) + ")");
+            }
+            do
+            {
+                stream.read((char*)in, sizeof(in));
+                strm.avail_in = (unsigned int)stream.gcount();
+                if (strm.avail_in == 0)
+                {
+                    break;
+                }
+                strm.next_in = in;
+                do
+                {
+                    strm.avail_out = ZLIB_CHUNK;
+                    strm.next_out = (Bytef*)out;
+                    ret = inflate(&strm, Z_NO_FLUSH);
+                    if (ret == Z_STREAM_ERROR)
+                    {
+                        XTENSOR_THROW(std::runtime_error, "zlib decompression failed (" + zlib_err(Z_STREAM_ERROR) + ")");
+                    }
+                    switch (ret) {
+                        case Z_NEED_DICT:
+                            ret = Z_DATA_ERROR;
+                        case Z_DATA_ERROR:
+                        case Z_MEM_ERROR:
+                            (void)inflateEnd(&strm);
+                            XTENSOR_THROW(std::runtime_error, "zlib decompression failed (" + zlib_err(ret) + ")");
+                    }
+                    have = ZLIB_CHUNK - strm.avail_out;
+                    // not possible to insert in a xt::svector currently
```

Comment:
It is, see https://github.com/xtensor-stack/xtensor/blob/c76e2936823a77a3beb7e5452e128cb6244d225c/include/xtensor/xstorage.hpp#L716